src/main/java/de/hysky/skyblocker/skyblock/end/TheEnd.java

```diff
@@ -0,0 +1,232 @@
+    );
+
+    public static ProtectorLocation currentProtectorLocation = null;
+    public static int stage = 0;
+
+    public static void init() {
+        AttackEntityCallback.EVENT.register((player, world, hand, entity, hitResult) -> {
+            if (entity instanceof EndermanEntity enderman && isZealot(enderman)) {
+                hitZealots.add(enderman.getUuid());
+            }
+            return ActionResult.PASS;
+        });
+
+
+        HudRenderCallback.EVENT.register((drawContext, tickDelta) -> {
+            if (!Utils.isInTheEnd()) return;
+            if (!SkyblockerConfigManager.get().locations.end.hudEnabled) return;
+
+            EndHudWidget.INSTANCE.render(drawContext, SkyblockerConfigManager.get().locations.end.enableBackground);
+        });
+
+        ClientChunkEvents.CHUNK_LOAD.register((world, chunk) -> {
+            String lowerCase = Utils.getIslandArea().toLowerCase();
+            if (Utils.isInTheEnd() || lowerCase.contains("the end") || lowerCase.contains("dragon's nest")) {
+                ChunkPos pos = chunk.getPos();
+                //
+                Box box = new Box(pos.getStartX(), 0, pos.getStartZ(), pos.getEndX(), 1, pos.getEndZ());
+                locationsLoop: for (ProtectorLocation protectorLocation : protectorLocations) {
+                    if (box.contains(protectorLocation.x, 0.5, protectorLocation.z)) {
+                        //MinecraftClient.getInstance().player.sendMessage(Text.literal("Checking: ").append(protectorLocation.name));//MinecraftClient.getInstance().player.sendMessage(Text.literal(pos.getStartX() + " " + pos.getStartZ() + " " + pos.getEndX() + " " + pos.getEndZ()));
+                        for (int i = 0; i < 5; i++) {
+                            if (world.getBlockState(new BlockPos(protectorLocation.x, i+5, protectorLocation.z)).isOf(Blocks.PLAYER_HEAD)) {
+                                stage = i+1;
+                                currentProtectorLocation = protectorLocation;
+                                EndHudWidget.INSTANCE.update();
+                                break locationsLoop;
+                            }
+                        }
+                    }
+                }
+                if (loaded.isEmpty()) load(); // Wacky fix for when you join skyblock, and you are directly in the end (profile id isn't parsed yet most of the time)
+            }
+
+
+        });
+        // Reset when changing island
+        // TODO: Replace when a changed island event is added
+        ClientPlayConnectionEvents.JOIN.register((handler, sender, client) -> {
+            resetLocation();
+            save();
+            load();
+        });
+        // Save when leaving as well
+        ClientLifecycleEvents.CLIENT_STOPPING.register((client) -> save());
+
+        ClientReceiveMessageEvents.GAME.register((message, overlay) -> {
+            String lowerCase = message.getString().toLowerCase();
```

Comment:
Add a check for when the player is in the end so that this isn't constantly ticked